### DISCUSSION TITLE
[ZK Stack] - How to properly deploy a hyperchain with the docker containers?

### DISCUSSION THREAD

**Discussion by ETeissonniere** [zkSync-Community-Hub/zksync-developers#281] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

stable-x86_64-unknown-linux-gnu

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

`zk stack docker-setup` does generate a working docker compose file but it seems unusable out of the box. The env file expects connections to a geth instance, postgres instance, etc... which are all not part of the docker compose file. Significant and undocumented modifications seem to be necessary to actually make use of the docker files.

### Expected Behavior

The docker compose file should be good enough to start a hyperchain via docker, or documentation should be provided.

### Current Behavior

Docker compose does not work out of the box, and necessary modifications are unclear. We have tried to start a local geth node, modify the docker compose file etc... but it appears that more details need to go into the setup as we were not able to complete the setup and had to revert to not using the docker images.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by SuccinctPaul** [comment]
Hi,guys, I've met the same problem and found it https://github.com/matter-labs/local-setup .

This repo provides a `docker-compose.yaml` and readme. 

Hope it can help u.

**Reply by ETeissonniere** [reply]
Thanks this is indeed helpful for a local setup. I will try to adapt it for a testnet sequencer and report back here

**Reply by dutterbutter** [reply]
Please see relevant documentation here: https://docs.zksync.io/zk-stack/running-a-hyperchain/production.html

**Reply by ETeissonniere** [reply]
@dutterbutter we already followed this. It does generate a docker compose file though it did not appear to work out of the box

**Reply by bxpana** [reply]
@ETeissonniere are you guys still having an issue with getting this to work?

**Reply by ETeissonniere** [reply]
@bxpana we were able to come up with a working docker compose configuration after a bit of fiddling around / patching the tooling. Happy to check if we can open it somewhere if relevant

**Reply by bxpana** [reply]
What kind of "fiddling around / patching the tooling" did you need to do?

**Reply by ETeissonniere** [reply]
Mainly reverse engineering the `zk` scripts and figuring out the right values to add to the config file. Of the top of my mind we met the following issues:
1. the `docker-setup` command did not create a docker compose file that works out of the box
2. containers were not tagged with the names the `zk docker-setup` script told us were used
3. even when we figured out what values to use our team had to disable the housekeeper service which led to no blocks being finalized on the testnet/Sepolia - we had to do some more reverse engineering to figure out how to enable it

The end result should be contained [here](https://gist.github.com/ETeissonniere/13a5fba888d09e18ef3394b853677614), though we since moved away from trying to host our own rollup sequencer for the time being so I can't guarantee this will quite work out of the box. I have also made an attempt to redact any sensitive values (look for `# REDACTED`) so you'd need to set those again.

**Reply by SuccinctPaul** [reply]
@bxpana  Would you provide a more details(from 0 to1) docs about how to deploy hyperchain in docker?


Besides, I have several questions: 
1. what's the difference between server-v2 and local-node and `zk-envoriment`  DockerFile?? 
2. How to deploy node and it's replica-one ?
3. How to deal when node shutdown on production environment.

**Reply by bxpana** [reply]
@SuccinctPaul in our docs you can run locally that uses docker https://docs.zksync.io/zk-stack/running-a-hyperchain/locally.html

Will check on your other questions.

**Reply by githubdoramon** [reply, marked as answer]
@ETeissonniere and @SuccinctPaul, the zk stack cli commands (actually, the whole zk set of commands) are being totally refactored now. Would it be ok for you to hold on a few weeks so we can get it out, or are you dealing with some time-sensitive tasks now?

**Reply by ETeissonniere** [reply]
yup - no worries

**Reply by SuccinctPaul** [reply]
Yeah, no problem for that.

**Reply by SuccinctPaul** [reply]
hi,@githubdoramon , any updates?

**Reply by githubdoramon** [reply]
We didn't finish the revamp yet, but now you can spin up local setup with ZK chains :) 

https://x.com/zkSyncDevs/status/1791172661554315639

**Reply by LimonChillo** [reply]
> @ETeissonniere and @SuccinctPaul, the zk stack cli commands (actually, the whole zk set of commands) are being totally refactored now. Would it be ok for you to hold on a few weeks so we can get it out, or are you dealing with some time-sensitive tasks now?

Hi @githubdoramon and @bxpana. I want to deploy my local zkStack into a containerized environment and have been looking through some threads here, this seemed to be the most fitting one. I saw that the `zk docker-setup` command is not yet available with the new `zkstack` CLI. What would be the best way to containerize my zkStack ecosystem as of today? Thanks for your help :)

**Reply by bxpana** [reply]
@LimonChillo before diving into this one just want to make sure that you are trying to do both, run ZK Stack in a docker container AND run on your Ubunutu server (https://github.com/zkSync-Community-Hub/zksync-developers/discussions/874)

**Reply by LimonChillo** [reply]
@bxpana Yes, I'm currently exploring both. Docker would be my preferred solution to run the zkStack alongside my other services. As I'm not sure about the right way to dockerize it, I'm also exploring the option to run the stack on a Ubuntu VM in the cloud.

**Reply by bxpana** [reply]
@LimonChillo this might be what you're looking for then, https://github.com/matter-labs/zksync-era/tree/f06cb79883bf320f50089099e0abeb95eaace470/docker/server-v2

There are also some [pre-built ones](https://hub.docker.com/r/matterlabs/server-v2/tags) where you just need to change the config files. Let me know if that helps.

**Comment by candacekrank** [comment]
It seems that the provided Docker setup for the zk stack lacks necessary configurations or documentation for seamless usability. To address this issue, detailed documentation should be provided alongside the Docker Compose file, explaining the required configurations and steps needed to set up additional components such as Geth and PostgreSQL. Alternatively, the Docker Compose file itself should be updated to include configurations for these components or to provide a more streamlined setup process.

**Comment by maggielyon** [comment]
It appears that the docker-compose file generated by zk stack requires additional configuration and modifications to work effectively. Consider reviewing the documentation or seeking assistance to ensure proper setup and usage of the dockerized environment.

**Comment by educatedb** [comment]
It seems that the zk stack docker setup lacks necessary configurations and documentation for seamless usability. It requires additional modifications and details to make it functional, hindering its out-of-the-box usability.

**Comment by colleennealv** [comment]
It seems there are issues with the zk stack docker-setup, as the generated docker compose file requires additional modifications to work properly. Documentation or improvements to the setup process may be needed to address this issue effectively.

**Comment by thakurmatt** [comment]
The user is encountering an issue with the zk stack docker-setup, as the generated docker compose file is not usable out of the box. The env file expects connections to a geth instance, postgres instance, etc. which are not part of the docker compose file. The necessary modifications are unclear, and the user has tried starting a local geth node and modifying the docker compose file without success. No repository link or additional details are provided.
